pkcs11-env.h

```diff
@@ -14,6 +14,9 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+/* From 2.1 of [PKCS11-base-v2.40]: Cryptoki structures SHALL be packed with 1-byte alignment. */
+#pragma pack(push, cryptoki, 1)
```

Comment:
Is this Windows-specific?  The [gcc docs](https://gcc.gnu.org/onlinedocs/gcc-8.1.0/gcc/Structure-Layout-Pragmas.html#Structure-Layout-Pragmas) imply a slightly different syntax, and the [Clang docs](http://releases.llvm.org/3.4/tools/clang/docs/UsersManual.html#microsoft-extensions) imply that `--fms-extensions` flag would be needed to cope with `#pragma push` of this form.

Reply:
I hadn't realized that, but it is Windows specific. [MSDN](https://msdn.microsoft.com/en-us/library/2e70t5y1.aspx) specifies the identifier as being optional, so perhaps GCC silently ignores it...

I removed the identifier as to make it comply with GCC's syntax (on commit 460feae). However, I'm not sure how to handle Clang (if at all)... 

The following test compiles as expected (and issues errors if any of the `#pragma`s is removed) without any special flag on GCC 4.8.4, Clang 3.6.1 and MinGW's GCC 5.4.0:

```c++
#include <cstddef> // includes offsetof
#include <cstdint> // includes uint32_t and uint8_t

struct Packed {
    uint32_t i1;
    uint8_t c;
    uint32_t i2;
};

#pragma pack(push, 1)
struct Unpacked {
    uint32_t i1;
    uint8_t c;
    uint32_t i2;
};
#pragma pack(pop)

struct PackedAfterPop {
    uint32_t i1;
    uint8_t c;
    uint32_t i2;
};

int main(int argc, char *argv[]) {
    // Check offsets of struct before push
    static_assert(offsetof(Packed, i1) == 0, "");
    static_assert(offsetof(Packed, c) == 4, "");
    static_assert(offsetof(Packed, i2) == 8, "");
    static_assert(sizeof(Packed) == 12, "");

    // Check offsets of packed struct
    static_assert(offsetof(Unpacked, i1) == 0, "");
    static_assert(offsetof(Unpacked, c) == 4, "");
    static_assert(offsetof(Unpacked, i2) == 5, "");
    static_assert(sizeof(Unpacked) == 9, "");

    // Check offsets of struct after pop
    static_assert(offsetof(PackedAfterPop, i1) == 0, "");
    static_assert(offsetof(PackedAfterPop, c) == 4, "");
    static_assert(offsetof(PackedAfterPop, i2) == 8, "");
    static_assert(sizeof(PackedAfterPop) == 12, "");

    return 0;
}
```

Reply:
Thank you for being so thorough!

One other worry -- if an existing PKCS#11 implementation *doesn't* do the packing, then linking the test code against that implementation won't work.  If that's the case, I guess we'd need an easier way of controlling "force-packed" or not (`#ifdef` maybe?).

I'll have a play with [opencryptoki](https://packages.debian.org/source/wheezy/opencryptoki) to see what happens there (and ideally add it to the Travis build), but it might be a few days/weeks before I can get to that.

Reply:
I'm not too happy with my current solution, but I couldn't come up with anything better: ad55b5d.

My main issue is requiring the user to supply a pre-defined, and somewhat arbitrary, variable when running `make`. In some of my more recent Makefiles, I've used `$MAKECMDGOALS`  and pseudo-rules (e.g., `make linux` or `make win`) to determine the value of variables, and whether they should be set at all. However, in this case, the latter seemed too complicated for merely `#define`'ing a macro, so I opted for using a variable.

Reply:
SGTM.

I did have a quick play with opencryptoki: it definitely needs the unpacked version of the headers (at least for the normal Debian Linux packaged version), so thanks for making the packing configurable.